corehq/apps/users/tasks.py

```diff
@@ -303,3 +306,33 @@ def update_domain_date(user_id, domain):
             and today > domain_membership.last_accessed):
         domain_membership.last_accessed = today
         user.save()
+
+
+@periodic_task(
+    run_every=deserialize_run_every_setting(settings.USER_REPORTING_METADATA_BATCH_SCHEDULE),
+    queue='background_queue',
+)
+def process_reporting_metadata_staging():
+    from corehq.apps.users.models import UserReportingMetadataStaging
+    from corehq.pillows.synclog import mark_last_synclog
+    from pillowtop.processors.form import mark_latest_submission
+
+    records = (
+        UserReportingMetadataStaging.objects.select_for_update(skip_locked=True).order_by('pk')
+    )[:100]
+    with transaction.atomic():
+        for record in records:
+            if record.received_on:
+                mark_latest_submission(
+                    record.domain, record.user_id, record.app_id, record.build_id,
+                    record.xform_version, record.form_meta, record.received_on
+                )
+            if record.device_id or record.sync_date:
+                mark_last_synclog(
```

Comment:
Currently these two functions can issue two saves to the user doc. This is not ideal and we could make a separate function to update all fields at once. I have not done this yet as this will keep it closer to the original code (and thus a bit easier to reason about these changes for now).